### ISSUE TITLE
Intersection tool creates duplicate vertices for line features

### ISSUE THREAD

**Issue by ejcorello** [qgis/QGIS#34934]
QGIS 3.10.3. The Geoprocessing Tool named Intersection creates duplicate vertices for all vertices that are not endpoints of the line segment. Consider input line layers A and B and the resulting Intersection line layer C. If v represents the number of vertices in the overlapping layer A or layer B segment, then the resulting intersecting line feature in layer C has 2v-2 vertices. There's one instance of each endpoint vertex, and 2 instances of every other vertex. For example, a line segment with 5 vertices in layer A or layer B has 8 vertices in layer C. When the lines are drawn, the layer C line can appear different than the layer A and layer B lines.

Maybe this condition also contributes to those times when I use the Difference tool, expecting a result of zero features, but instead I get a couple of lines with only 2 vertices. Regardless of whether I use A-B or B-A.

**Comment by saberraz** [comment]
- what file format are you using?
- if it is a file based layer, is it local or shared folder on a network drive?

I have seen reports of similar issue for shapefile on Windows network drive.

**Comment by ejcorello** [comment]
I'm using a .shp shapefile.  It is a local file, not shared, residing on my own internal hard drive.

**Comment by gioman** [comment]
@ejcorello please attach sample data (also if possible attach obtained results VS expected).

**Comment by gioman** [comment]
@ejcorello and by the way: is this about the native QGIS intersection tool?

**Comment by ejcorello** [comment]
Yes this is the native Intersection tool found under Vector, Geoprocessing Tools, Intersection...
The result is found in Intersection layer C. The Vertex Editor shows the 8 vertices. The expected result is that the Vertex Editor for layer C shows only the same 5 vertices that are shown for layer A or layer B.
[Intersection test.zip](https://github.com/qgis/QGIS/files/4303795/Intersection.test.zip)

**Comment by gioman** [comment]
>  The Vertex Editor shows the 8 vertices

@ejcorello the results it a multipart layer, if you make it singlepart you will see that is made of 4 lines/features, so the 8 vertices are correct.

Maybe here the reall question if is expected this tool to return multipart in this specific scenario (as the inputs are not multipart): the output is really made of 4 separate lines while inputs are not.

**Comment by ejcorello** [comment]
Giovanni,

Makes sense. How did you determine that it's a multipart feature?

I guess maybe I need to find a tool that will convert a multipart feature to a single singlepart feature. Do you know of such a tool? I see that there's a tool called Multipart to singleparts but I don't want to create a new feature for each part. 

I want to be able to use the same symbology for the input layers and also for the intersected layer C and that doesn't seem to work very well when mixing single part and multipart. 

Eric
     On Sunday, March 8, 2020, 05:14:07 PM EDT, Giovanni Manghi <notifications@github.com> wrote:  
 
 

The Vertex Editor shows the 8 vertices


@ejcorello the results it a multipart layer, if you make it singlepart you will see that is made of 4 lines/features, so the 8 vertices are correct.

Maybe here the reall question if is expected this tool to return multipart in this specific scenario (as the inputs are not multipart): the output is really made of 4 separate lines while inputs are not.

**Comment by gioman** [comment]
> How did you determine that it's a multipart feature?

@ejcorello by using the tool "multipart to singleparts"



> I want to be able to use the same symbology for the input layers and also for the intersected layer C and that doesn't seem to work very well when mixing single part and multipart.

I don't think that it makes any difference in terms of symbology.


The real question remains: with the provided inputs is expected to get a multipart output from the intersection tool?

**Comment by ejcorello** [comment]
With all respect, I think that it does make a difference in terms of symbology. I will try things again to be sure.

The only reason I noticed this issue in the first place is because I could see the visible difference between the input features and the multipart output features when they are rendered, using much larger layers with many more features and many more intersecting and non-intersecting segments with many many more vertices. Of course this is after I use the Difference tool to remove the intersecting segments from the input layers, leaving me with three mutually exclusive layers.

You are right about the underlying question. Maybe I should open a feature enhancement request or whatever it's called after I've verified the visible difference again.

     On Sunday, March 8, 2020, 06:14:39 PM EDT, Giovanni Manghi <notifications@github.com> wrote:  
 
 

How did you determine that it's a multipart feature?


@ejcorello by using the tool "multipart to singleparts"


I want to be able to use the same symbology for the input layers and also for the intersected layer C and that doesn't seem to work very well when mixing single part and multipart.


I don't think that it makes any difference in terms of symbology.

The real question remains: with the provided inputs is expected to get a multipart output from the intersection tool?

**Comment by gioman** [comment]
>  I think that it does make a difference in terms of symbology

@ejcorello can you make an example?

**Comment by ejcorello** [comment]
Yes I can make an example but first I will say that I think I found a solution to the underlying issue that the input layers are single part but the resultant Intersection layer is multipart. The solution is simply to use the "Merge lines" tool on the multipart Intersection layer. It "joins all connected parts of MultiLineString geometries into single LineString geometries." I've tried it. It appears to be just what I'm looking for and I will work under that assumption for now and continue with my project.

In the attached example you can see that the cross-hatching on the right and left sides is regularly spaced but the cross-hatching in the middle is not spaced the same way. If you replace the "Intersection A x B multipart" layer with "Merged Intersection" layer, created with the Merge Lines tool, then the cross-hatching in the middle is the same as the left and right sides.
     On Monday, March 9, 2020, 07:49:45 AM EDT, Giovanni Manghi <notifications@github.com> wrote:  
 
 

I think that it does make a difference in terms of symbology


@ejcorello can you make an example?

**Comment by gioman** [comment]
> Yes I can make an example

@ejcorello do you mean this?

![Screenshot_20200309_150746](https://user-images.githubusercontent.com/1951107/76227885-e12fd680-6217-11ea-9f04-6ae112e47032.png)

then you get around that with "symbol levels"

![Screenshot_20200309_150808](https://user-images.githubusercontent.com/1951107/76227950-f60c6a00-6217-11ea-88e0-fbf5687a0fe1.png)

![Screenshot_20200309_150821](https://user-images.githubusercontent.com/1951107/76227960-f9075a80-6217-11ea-9e5d-4643de40885d.png)

**Comment by gioman** [comment]
> In the attached example

no new attachment.

**Comment by ejcorello** [comment]
I attached it to an email. I'll drag and drop it here. OK? 
[symbology test.zip](https://github.com/qgis/QGIS/files/4307906/symbology.test.zip)

**Comment by gioman** [comment]
> I attached it to an email. I'll drag and drop it here. OK?

ok, but this is not anyway the subject of this ticket. So let's back on track. The main question still is:
with the provided inputs is expected to get a multipart output from the intersection tool?

**Comment by alexbruy** [comment]
@gioman this (multipart output) is expected and correct behaviour. We had discussion about output type for overlay tools and agreed that intersection and some other tools should always output multi-geometry.

**Comment by ejcorello** [comment]
![symbolgy2](https://user-images.githubusercontent.com/61944715/76351407-c75bc580-62e3-11ea-8e6c-3f77a9304163.jpg)
@alexbruy  It is what I suspected, that multipart output is correct, given the many possible combinations of singlepart and multipart inputs. For a novice user like me, who was clueless about singlepart vs. multipart before I encountered this behavior, maybe there could be a sentence in the tool about the output geometry, similar to the sentence in the Dissolve tool.
